### PR TITLE
57 distributed lock on db schema patch

### PR DESCRIPTION
**Pull request by akhtyamovrr**

### CODE REVIEW

File: storages/factory.go [hunk 1]
```diff
@@ -54,6 +55,12 @@ func Create(ctx context.Context, destinations *viper.Viper, logEventPath string)
 		return stores, consumers
 	}
 
+	monitorKeeper, err := NewMonitorKeeper(syncServiceType, syncServiceEndpoint, connectionTimeoutSeconds)
```
Comment (xtreding) [inline storages/factory.go:58]: MonitorKeeper is an important part of EventNative and if we have an error we shouldn't start the server.
Please use log.Fatal() for this case. (after moving to main.go)

File: storages/monitor_keeper.go [hunk 2]
```diff
@@ -1,32 +1,119 @@
...
+
+func (emk *EtcdMonitorKeeper) Unlock(lock Lock) error {
+	ctx := context.Background()
+	if err := lock.Unlock(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (emk *EtcdMonitorKeeper) GetVersion(dbType string, tableName string) (int64, error) {
```
Comment (xtreding) [inline storages/monitor_keeper.go:70]: dbType isn't unique.
e.g. service can have two postgres destinations.
Use destinationName instead

**Review by xtreding — CHANGES_REQUESTED** [review]

File: appconfig/appconfig.go [hunk 3]
```diff
@@ -22,6 +22,11 @@ type AppConfig struct {
 	GeoResolver geo.Resolver
 	UaResolver  useragent.Resolver
 
+	SynchronizationServiceType       string
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply appconfig/appconfig.go:25]: moved sync service setting processing to main

File: storages/monitor_keeper.go [hunk 4]
```diff
@@ -1,32 +1,119 @@
...
+			Endpoints:   []string{syncServiceEndpoint},
+		})
+		if err != nil {
+			return nil, err
+		}
+		keeper := &EtcdMonitorKeeper{}
+		keeper.client = cli
+		return keeper, nil
+	} else {
+		return nil, errors.Unwrap(fmt.Errorf("Unknown sync service type %s.", syncServiceType))
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply storages/monitor_keeper.go:117]: You are right

File: storages/monitor_keeper.go [hunk 5]
```diff
@@ -1,32 +1,119 @@
...
+		cli, err := clientv3.New(clientv3.Config{
+			DialTimeout: time.Duration(connectionTimeoutSeconds) * time.Second,
+			Endpoints:   []string{syncServiceEndpoint},
+		})
+		if err != nil {
+			return nil, err
+		}
+		keeper := &EtcdMonitorKeeper{}
+		keeper.client = cli
+		return keeper, nil
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply storages/monitor_keeper.go:115]: done

File: storages/factory.go (hunk 1 above)
Comment (akhtyamovrr) [reply storages/factory.go:58]: done

File: storages/monitor_keeper.go [hunk 6]
```diff
@@ -1,32 +1,119 @@
...
+	ctx := context.Background()
+	if err := l.Lock(ctx); err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+func (emk *EtcdMonitorKeeper) Unlock(lock Lock) error {
+	ctx := context.Background()
+	if err := lock.Unlock(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply storages/monitor_keeper.go:64]: I've added closer processing so unlock is changed. It closes the resource only if unlock is successful as this function may be called in retry

File: storages/monitor_keeper.go [hunk 7]
```diff
@@ -1,32 +1,119 @@
...
+	ctx := context.Background()
+	kv := emk.client
+	version = version + 1
+	_, putErr := kv.Put(ctx, dbType+"_"+tableName, strconv.FormatInt(version, 10))
+	return version, putErr
+}
+
+func NewMonitorKeeper(syncServiceType string, syncServiceEndpoint string, connectionTimeoutSeconds uint) (MonitorKeeper, error) {
+	if syncServiceType == "" || syncServiceEndpoint == "" {
+		fmt.Println("Using stub sync server as no configuration is provided")
```
[the remark this answers is not in the document]
Comment (akhtyamovrr) [reply storages/monitor_keeper.go:102]: fixed

File: storages/monitor_keeper.go (hunk 2 above)
Comment (akhtyamovrr) [reply storages/monitor_keeper.go]: fixed

File: main.go [hunk 8]
```diff
@@ -118,11 +118,19 @@ func main() {
 		appconfig.Instance.ScheduleClosing(logger)
 	}
 
+	syncServiceType := viper.GetString("synchronization_service.type")
+	syncServiceEndpoint := viper.GetString("synchronization_service.endpoint")
+	viper.SetDefault("synchronization_service.connection_timeout_seconds", 20)
```
Comment (xtreding) [inline main.go:123]: It would be better to set default values in one place (appconfig.setDefaultParams())

**Review by xtreding — APPROVED** [review]
Please add information about the new version in CHANGELOG.md

File: main.go (hunk 8 above)
Comment (akhtyamovrr) [reply main.go:123]: Thanks, fixed

**Pull request merged**